### ISSUE TITLE
windows: SGR-1 incorrectly resets the color

### ISSUE THREAD

**Issue by avih** [gwsw/less#377]
If invoking `less -R` and there's an SGR sequence of some color, followed by SGR sequence of bold/bright, then the color is reset and only bright is displayed.

The other way around (bright, then some color) does work correctly, i.e. it shows the color in bright.

For instance this (using [busybox-w32](https://github.com/rmyorston/busybox-w32) for `printf`):
```sh
busybox printf "\033[32m green \033[1m bright \033[m reset \033[1m bright \033[32m green \033[0m\n" | .\less.exe -RXE
```

prints SGR32 green SGR1 bright SGR0 reset SGR1 bright SGR32 green, which should display as normal-green, bright-green, plain-white, bright-white, bright-green.

However, in practice the second text is displayed as bright-only (bright cyan actually, as that's the default bold color of `less` on Windows), instead of bright green.

Notice that the other way around (the last "green" which is after SGR1 followed by SGR32) is correctly bright-green.

Tested with latest official release for windows ([v633](https://github.com/jftuga/less-Windows/releases/tag/less-v633)) but also with my own builds, and this issue exists since at least v540 (in 2018).

Screenshot in `cmd.exe` window:
![less-color-bright-win32](https://github.com/gwsw/less/assets/2164962/7357be48-c809-4984-b678-ff5992b49b9a)

EDIT: interestingly, green followed by bright at the same SGR sequence does show correctly, e.g. `printf "\033[32;1m bright-green \033[m" | less -R` does show as bright-green.

**Comment by gwsw** [comment]
I haven't looked into this in detail yet, but does it behave differently if you use the ```-Da``` option?

**Comment by avih** [comment]
> does it behave differently if you use the `-Da` option?

Yes, adding `-Da` works around the issue and displays correctly.

**Comment by gwsw** [comment]
I'm glad to hear that. The documentation of -Da seems to have been accidentally removed from the man page at some point. I will add it back.

**Comment by gwsw** [comment]
Documented in man page in 9fb64a3120f8c196ecb03b8b850502edc3227e65.

**Comment by avih** [comment]
> On MS-DOS only, the \-Da option may be used to specify strict parsing of ANSI color (SGR) sequences when the \-R option is used.Without this option, sequences that change text attributes (bold, underline, etc.) may clear the text color.

There are not many MS DOS users currently, I'd imagine, but there are many Windows users which are probably affected.

What does "may" mean? depending on what? What's the difference between strict and non strict parsing? Why is this not enabled by default, and why does this option exist?

**Comment by avih** [comment]
OK, after some experiments, here are my understanding so far. Do correct me if I'm wrong.

On windows 10, `-Da` enables "plain ansi" mode, and combined with `vt_mode` (which gets enabled automatically where supported, i.e. win 10+), ends up not modifying the sequences at all. I.e. It lets the terminal handle all the escape sequences, including 256 and true colors. So that's as perfect as it can get IMHO.

On earlier Windows versions, like 8 or 7, where VT mode doesn't exist, if `-Da` is set, then `less` emulates an ANSI terminal (specifically, the "basic" SGR sequences) as best as it can.

Since not all attributes are supported by such consoles, `less` emulates them (inverse), or maps them to something else (italic, underline, conceal), or ignores them (crossed-out, dim).

Specifically, italic seems to be mapped to inverse, conceal to bold-black fg on black bg, and underline to bold-black bg (without touching the fg).

However, this arbitrary mapping (or any other) can end up non-legible in some cases, and that's where the "without `-Da`" mode kicks in (i.e. the default).

Without `-Da` it's the same like with `-Da`, except that if a char has default fg and default bg and exactly one attribute of bold/inverse/underline/italic, then it's mapped to a specific other color (fg and or bg, each possibly bright), as configurable via the other `-D` modes.

So adding `-Da` simply disables this default-color-with-exactly-one-attribute translation.

Here's a screenshot on Windows 7. It uses [sgrtab](https://github.com/avih/sgrtab) to generate the colors.

The first block is how busybox-w32 emulates it (it basically ignores all attributes except bold and inverse).

The second block is default `less -R` (i.e. without `-Da`) where it emulates the terminal generically but also translates default-color-with-one-attribute (left column).

The third block is with `-Da`, where it's "generic" terminal emulation without exception.
![less-sgrtab-info-win7](https://github.com/gwsw/less/assets/2164962/97ce9bee-3007-49b3-a082-41cd9296b4b1)

**Comment by gwsw** [comment]
@adoxa Hi Jason, does avih's description above sound correct to you? When you originally wrote this sgr_mode code, you described it like this:

>The default
>behavior is to use the assigned colors for individual attributes (this is
>how I think less.man should be displayed).  A new option -Da will force
>strict SGR interpretation, displaying more like ANSICON (where underline
>and blink use bright background).  In this mode, less.man doesn't display
>right if the normal foreground is already bright (such as mine is).  It
>should, however, work better with other ANSI files.

**Comment by avih** [comment]
> The default
> behavior is to use the assigned colors for individual attributes (this is
> how I think less.man should be displayed). A new option -Da will force
> strict SGR interpretation, displaying more like ANSICON (where underline
> and blink use bright background). In this mode, less.man doesn't display
> right if the normal foreground is already bright (such as mine is). It
> should, however, work better with other ANSI files.

For what it's worth, my estimation is that the main issue was how underline looked before this change (at the bottom of my screenshot - with `-Da`), and since underline is used in man pages, it's useful to be able to display it better.

The other attributes (bold, italic) actually look fine with `-Da` (i.e. generic emulation without additional translation), so I wouldn't think they'd need to be mapped to something else.

If I had to solve it back then, I'd just change underline to be the same as bold (or italic the same as bold, and underline to inverse).

Because the console does support bold/inverse(logically) great, so translating unsupported attributes to existing supported attribute would look fine.

So basically underline didn't look good, so make the specific bad thing (underline in manpages) look OK. All the rest is just collateral damage IMHO.

**Comment by adoxa** [comment]
I'm not entirely sure what the issue has ended up being. My view regarding man pages is if you want a better underline it's up to you to define `-Du` (and leave `-Da` off). On Windows 10/11 a real underline is possible with `-Duu`.

**Comment by avih** [comment]
I think at this stage the first question is how to document `-Da`.

The current docs, which was pushed few days ago, is:
> On MS-DOS only, the -Da option may be used to specify strict parsing of ANSI color (SGR) sequences when the -R option is used.Without this option, sequences that change text attributes (bold, underline, etc.) may clear the text color.

With which I have 2 gripes:
1. At the very least it also applies to Windows users (one could guess primarily to windows users), so it should be stated.
2. It suggests that `-Da` is more rigorous or strict than the default mode (without `-Da`), But I think it's incorrect. I think that with `-Da` is standard ANSI interpretation, with programatic, non-configurable mappings of SGR attributes to something else (e.g. italic to inverse), while without `-Da` is a non-standard interpretation of SGR sequences, where as long as an SGR sequence doesn't contain exactly one attribute - it's identical to with `-Da`, but an SGR sequence with a single attribute also non-standardly resets the colors to the ones configured for this attribute by the `-D{d,k,s,u}...` (or the default colors for these options).

So this `\e[33;1m text` SGR sequence is not a single attribute on its own, hence interpreted the same regardless of `-Da`, and produces bright-yellow text.

But this `\e[33m\e[1m text` has two SGR sequences where the second is of a single attribute.

`-Da` will interpret it in a standard way, i.e. yellow+bright transformed however the programatic transformation does (in this case not much - bold/bright is supported by the console), so it ends up bright yellow too.

But without `-Da`, because the second SGR is a sole attribute, it resets the color to whatever was configured with `-Dd...` (or its default, which happens to be bright cyan).

So the first question is whether you agree with the above.

If we agree on that, we might go further:

Why was `-Da` added in the first place, and is there a better solution.

My estimation was that it was added *mainly* because underline specifically is not very legible with the default, non-configurable, programatic translation of the underline attribute - which brightens the BG color.

As seen by the bottom block of my screenshot above (with `-Da`), the programatic translation of underline is indeed not very legible, and so the default behavior became non-standard with this additional configurable translation, and the old, standard behavior was preserved with the `-Da` option.

If that's correct, then we can maybe try to find better or different solution, which preferably will keep standard interpretation by default, instead of under the `-Da` option.

Such solution might include, for instance, a different programatic translation of the underline attribute.

**Comment by adoxa** [comment]
`-Da` was added because I wanted the single-attribute sequences to use the defined colours (for man pages), but also a way to use standard sequences (for ANS(I) files, using ANSI.SYS/ANSICON or now Windows 10/11 itself). I prefer the defined colours because my default colour is already bright, so the single bold sequence had no effect and man pages didn't look right.

Underline sets bright background, because what is else is it going to do? It's that or nothing (when not on Win10).

**Comment by avih** [comment]
Can we go one step at a time?

Do you agree with my comments about the docs?

> I prefer the defined colours because my default colour is already bright, so the single bold sequence had no effect

That's non default, right? i.e. for most people which didn't change the terminal default (but I'm guessing also for many who did change them), bold with `-Da` will look brighter than without bold, right? (as my screenshot happens to show in my configuration).

So at least for bold/bright, I'm guessing the vast majority of users will see a distinct difference between bright and non-bright even with `-Da`.

Notice that I'm not arguing that allowing custom config is bad, just that the non-configurable (with `-Da`) would still be fine for most users, at least for bold/bright.

> Underline sets bright background, because what is else is it going to do? It's that or nothing

We're not there yet IMO, but It can do many things. For instance get mapped to bold. or to inverse (like italics), or to bold+inverse, or whatever.

"underline brightens the bg" is definitely not the only concievable mapping for underline IMHO.

**Comment by adoxa** [comment]
I think the doc is reasonable enough. It used to be that MS-DOS would imply Windows, but I guess time has moved on...

Sure, my colour is non-standard, but I still think defining the colours and using those is the better default for Windows (which implies MS-DOS :) ).

Whatever less does for underline is going to be sub-optimal, I'd rather leave it as is, more-or-less distinct from other attributes.

I'm still not sure what you actually want? `-Da` to become the default?

**Comment by avih** [comment]
> I'm still not sure what you actually want? `-Da` to become the default?

For starter, to not document that `-Da` is strict.

The fact is that `-Da` is the standard interpretation of SGR sequences, while without `-Da` is a non standard interpretation. Not only because it's configurable, but because it incorrectly resets the colors when it encounters an SGR sequence which is composed of a sole attribute.

**Comment by 1268** [comment]
https://github.com/martinvonz/jj/issues/1930

**Comment by 1268** [comment]
https://github.com/gwsw/less/issues/410

**Comment by avih** [comment]
Note that there's #383 to improve the 8/16 colors emulation states, which might fix the diff color issue.

As for 8 bit colors, that indeed can be enabled with `-Da` - which bypasses the internal `less` windows color emulation, and lets the terminal/console handle it.

On windows 10+ the console indeed supports true color (24 bit RGB), and `-Da` works. On earlier versions of windows `-Da` won't work.

**Comment by 1268** [comment]
@avih thanks for the update. I hope the referenced PR can be merged soon, and that this issue can be closed as well. running into this issue was quite upsetting, as it completely derailed what I was actually working on, and forced me to troubleshoot my Terminal and Shell, only to find out the issue was actually with the pager. Until this issue is closed I am going to be recommending other pagers such as https://github.com/tigrawap/slit for Windows users.

a pager is a widely used "dependency", so its important that it "just work" out of the box, ideally without having to futz with many configuration settings.

**Comment by garoto** [comment]
The guy is probably on his third GH account by now, but still can't help himself about deleting his own posts. Just too precious.

**Comment by 1268** [comment]
looks like we found the GitHub troll, with his whopping 0 repositories:

https://github.com/garoto?tab=repositories&type=source

**Comment by garoto** [comment]
It's kinda funny how you try to frame me as some sort of a person whose's opinion shoudn't be considered, by using the metric of my GH OG repo count, precious yet again. You know the news involving your garbage nick will be "`89z` shoots 12 in a shopping mall in fricking Indiana" in the not so distant future right? Because you're beyond help by now.
